Libc/gen/sysctl.3
.\" Copyright (c) 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)sysctl.3	8.4 (Berkeley) 5/9/95
.\" $FreeBSD: src/lib/libc/gen/sysctl.3,v 1.63 2004/07/02 23:52:10 ru Exp $
.\"
.Dd October 21, 2008
.Dt SYSCTL 3
.Os
.Sh NAME
.Nm sysctl ,
.Nm sysctlbyname ,
.Nm sysctlnametomib
.Nd get or set system information
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In sys/types.h
.In sys/sysctl.h
.Ft int
.Fn sysctl "int *name" "u_int namelen" "void *oldp" "size_t *oldlenp" "void *newp" "size_t newlen"
.Ft int
.Fn sysctlbyname "const char *name" "void *oldp" "size_t *oldlenp" "void *newp" "size_t newlen"
.Ft int
.Fn sysctlnametomib "const char *name" "int *mibp" "size_t *sizep"
.Sh DESCRIPTION
The
.Fn sysctl
function retrieves system information and allows processes with
appropriate privileges to set system information.
The information available from
.Fn sysctl
consists of integers, strings, and tables.
Information may be retrieved and set from the command interface
using the
.Xr sysctl 8
utility.
.Pp
Unless explicitly noted below,
.Fn sysctl
returns a consistent snapshot of the data requested.
Consistency is obtained by locking the destination
buffer into memory so that the data may be copied out without blocking.
Calls to
.Fn sysctl
are serialized to avoid deadlock.
.Pp
The state is described using a ``Management Information Base'' (MIB)
style name, listed in
.Fa name ,
which is a
.Fa namelen
length array of integers.
.Pp
The
.Fn sysctlbyname
function accepts an ASCII representation of the name and internally
looks up the integer name vector.
Apart from that, it behaves the same
as the standard
.Fn sysctl
function.
.Pp
The information is copied into the buffer specified by
.Fa oldp .
The size of the buffer is given by the location specified by
.Fa oldlenp
before the call,
and that location gives the amount of data copied after a successful call
and after a call that returns with the error code
.Er ENOMEM .
If the amount of data available is greater
than the size of the buffer supplied,
the call supplies as much data as fits in the buffer provided
and returns with the error code
.Er ENOMEM .
If the old value is not desired,
.Fa oldp
and
.Fa oldlenp
should be set to NULL.
.Pp
The size of the available data can be determined by calling
.Fn sysctl
with the
.Dv NULL
argument for
.Fa oldp .
The size of the available data will be returned in the location pointed to by
.Fa oldlenp .
For some operations, the amount of space may change often.
For these operations,
the system attempts to round up so that the returned size is
large enough for a call to return the data shortly thereafter.
.Pp
To set a new value,
.Fa newp
is set to point to a buffer of length
.Fa newlen
from which the requested value is to be taken.
If a new value is not to be set,
.Fa newp
should be set to NULL and
.Fa newlen
set to 0.
.Pp
The
.Fn sysctlnametomib
function accepts an ASCII representation of the name,
looks up the integer name vector,
and returns the numeric representation in the mib array pointed to by
.Fa mibp .
The number of elements in the mib array is given by the location specified by
.Fa sizep
before the call,
and that location gives the number of entries copied after a successful call.
The resulting
.Fa mib
and
.Fa size
may be used in subsequent
.Fn sysctl
calls to get the data associated with the requested ASCII name.
This interface is intended for use by applications that want to
repeatedly request the same variable (the
.Fn sysctl
function runs in about a third the time as the same request made via the
.Fn sysctlbyname
function).
The
.Fn sysctlnametomib
function is also useful for fetching mib prefixes and then adding
a final component.
For example, to fetch process information
for processes with pid's less than 100:
.Pp
.Bd -literal -offset indent -compact
int i, mib[4];
size_t len;
struct kinfo_proc kp;

/* Fill out the first three components of the mib */
len = 4;
sysctlnametomib("kern.proc.pid", mib, &len);

/* Fetch and print entries for pid's < 100 */
for (i = 0; i < 100; i++) {
	mib[3] = i;
	len = sizeof(kp);
	if (sysctl(mib, 4, &kp, &len, NULL, 0) == -1)
		perror("sysctl");
	else if (len > 0)
		printkproc(&kp);
}
.Ed
.Pp
Note:  Implementation of
.Fn printkproc
-- to print whatever data deemed necessary from the large
.Vt kinfo_proc
structure (
.In sysctl.h
) -- is left as an exercise for the reader.
.Pp
The top level names are defined with a CTL_ prefix in
.In sys/sysctl.h ,
and are as follows.
The next and subsequent levels down are found in the include files
listed here, and described in separate sections below.
.Pp
.Bl -column CTLXMACHDEP "Next level names" -offset indent
.It Sy "Name	Next level names	Description"
.It "CTL_DEBUG	sys/sysctl.h	Debugging"
.It "CTL_VFS	sys/mount.h	File system"
.It "CTL_HW	sys/sysctl.h	Generic CPU, I/O"
.It "CTL_KERN	sys/sysctl.h	High kernel limits"
.It "CTL_MACHDEP	sys/sysctl.h	Machine dependent"
.It "CTL_NET	sys/socket.h	Networking"
.It "CTL_USER	sys/sysctl.h	User-level"
.It "CTL_VM	sys/resources.h	Virtual memory (struct loadavg)"
.It "CTL_VM	sys/vmmeter.h	Virtual memory (struct vmtotal)"
.El
.Pp
For example, the following retrieves the maximum number of processes allowed
in the system:
.Pp
.Bd -literal -offset indent -compact
int mib[2], maxproc;
size_t len;

mib[0] = CTL_KERN;
mib[1] = KERN_MAXPROC;
len = sizeof(maxproc);
sysctl(mib, 2, &maxproc, &len, NULL, 0);
.Ed
.Pp
To retrieve the standard search path for the system utilities:
.Pp
.Bd -literal -offset indent -compact
int mib[2];
size_t len;
char *p;

mib[0] = CTL_USER;
mib[1] = USER_CS_PATH;
sysctl(mib, 2, NULL, &len, NULL, 0);
p = malloc(len);
sysctl(mib, 2, p, &len, NULL, 0);
.Ed
.Ss CTL_DEBUG
The debugging variables vary from system to system.
A debugging variable may be added or deleted without need to recompile
.Fn sysctl
to know about it.
Each time it runs,
.Fn sysctl
gets the list of debugging variables from the kernel and
displays their current values.
The system defines twenty
.Pq Vt "struct ctldebug"
variables named
.Va debug0
through
.Va debug19 .
They are declared as separate variables so that they can be
individually initialized at the location of their associated variable.
The loader prevents multiple use of the same variable by issuing errors
if a variable is initialized in more than one place.
For example, to export the variable
.Va dospecialcheck
as a debugging variable, the following declaration would be used:
.Pp
.Bd -literal -offset indent -compact
int dospecialcheck = 1;
struct ctldebug debug5 = { "dospecialcheck", &dospecialcheck };
.Ed
.Ss CTL_VFS
A distinguished second level name, VFS_GENERIC,
is used to get general information about all file systems.
One of its third level identifiers is VFS_MAXTYPENUM
that gives the highest valid file system type number.
Its other third level identifier is VFS_CONF that
returns configuration information about the file system
type given as a fourth level identifier (see
.Xr getvfsbyname 3
as an example of its use).
The remaining second level identifiers are the
file system type number returned by a
.Xr statfs 2
call or from VFS_CONF.
The third level identifiers available for each file system
are given in the header file that defines the mount
argument structure for that file system.
.Ss CTL_HW
The string and integer information available for the CTL_HW level
is detailed below.
The changeable column shows whether a process with appropriate
privilege may change the value.
.Bl -column "Second level nameXXXXXX" integerXXX -offset indent
.It Sy "Second level name	Type	Changeable"
.It "HW_MACHINE	string	no"
.It "HW_MODEL	string	no"
.It "HW_NCPU	integer	no"
.It "HW_BYTEORDER	integer	no"
.It "HW_PHYSMEM	integer	no"
.It "HW_MEMSIZE	integer	no"
.It "HW_USERMEM	integer	no"
.It "HW_PAGESIZE	integer	no"
.It "HW_FLOATINGPOINT	integer	no"
.It "HW_MACHINE_ARCH	string	no"
.\".It "HW_DISKNAMES	integer	no"
.\".It "HW_DISKSTATS	integer	no"
.El
.Pp
.Bl -tag -width 6n
.It Li HW_MACHINE
The machine class.
.It Li HW_MODEL
The machine model
.It Li HW_NCPU
The number of cpus.
.It Li HW_BYTEORDER
The byteorder (4,321, or 1,234).
.It Li HW_PHYSMEM
The bytes of physical memory represented by a 32-bit integer (for backward compatibility). Use HW_MEMSIZE instead.
.It Li HW_MEMSIZE
The bytes of physical memory represented by a 64-bit integer.
.It Li HW_USERMEM
The bytes of non-kernel memory.
.It Li HW_PAGESIZE
The software page size.
.It Li HW_FLOATINGPOINT
Nonzero if the floating point support is in hardware.
.It Li HW_MACHINE_ARCH
The machine dependent architecture type.
.\".It Fa HW_DISKNAMES
.\".It Fa HW_DISKSTATS
.El
.Ss CTL_KERN
The string and integer information available for the CTL_KERN level
is detailed below.
The changeable column shows whether a process with appropriate
privilege may change the value.
The types of data currently available are process information,
system vnodes, the open file entries, routing table entries,
virtual memory statistics, load average history, and clock rate
information.
.Bl -column "KERNXMAXFILESPERPROCXXX" "struct clockrateXXX" -offset indent
.It Sy "Second level name	Type	Changeable"
.It "KERN_ARGMAX	integer	no"
.It "KERN_BOOTFILE	string	yes"
.It "KERN_BOOTTIME	struct timeval	no"
.It "KERN_CLOCKRATE	struct clockinfo	no"
.It "KERN_FILE	struct file	no"
.It "KERN_HOSTID	integer	yes"
.It "KERN_HOSTNAME	string	yes"
.It "KERN_JOB_CONTROL	integer	no"
.It "KERN_MAXFILES	integer	yes"
.It "KERN_MAXFILESPERPROC	integer	yes"
.It "KERN_MAXPROC	integer	no"
.It "KERN_MAXPROCPERUID	integer	yes"
.It "KERN_MAXVNODES	integer	yes"
.It "KERN_NGROUPS	integer	no"
.It "KERN_NISDOMAINNAME	string	yes"
.It "KERN_OSRELDATE	integer	no"
.It "KERN_OSRELEASE	string	no"
.It "KERN_OSREV	integer	no"
.It "KERN_OSTYPE	string	no"
.It "KERN_POSIX1	integer	no"
.It "KERN_PROC	struct proc	no"
.It "KERN_PROF	node	not applicable"
.It "KERN_QUANTUM	integer	yes"
.It "KERN_SAVED_IDS	integer	no"
.It "KERN_SECURELVL	integer	raise only"
.It "KERN_UPDATEINTERVAL	integer	no"
.It "KERN_VERSION	string	no"
.It "KERN_VNODE	struct vnode	no"
.El
.Pp
.Bl -tag -width 6n
.It Li KERN_ARGMAX
The maximum bytes of argument to
.Xr execve 2 .
.It Li KERN_BOOTFILE
The full pathname of the file from which the kernel was loaded.
.It Li KERN_BOOTTIME
A
.Va struct timeval
structure is returned.
This structure contains the time that the system was booted.
.It Li KERN_CLOCKRATE
A
.Va struct clockinfo
structure is returned.
This structure contains the clock, statistics clock and profiling clock
frequencies, the number of micro-seconds per hz tick and the skew rate.
.It Li KERN_FILE
Return the entire file table.
The returned data consists of a single
.Va struct filehead
followed by an array of
.Va struct file ,
whose size depends on the current number of such objects in the system.
.It Li KERN_HOSTID
Get or set the host id.
.It Li KERN_HOSTNAME
Get or set the hostname.
.It Li KERN_JOB_CONTROL
Return 1 if job control is available on this system, otherwise 0.
.It Li KERN_MAXFILES
The maximum number of files that may be open in the system.
.It Li KERN_MAXFILESPERPROC
The maximum number of files that may be open for a single process.
This limit only applies to processes with an effective uid of nonzero
at the time of the open request.
Files that have already been opened are not affected if the limit
or the effective uid is changed.
.It Li KERN_MAXPROC
The maximum number of concurrent processes the system will allow.
.It Li KERN_MAXPROCPERUID
The maximum number of concurrent processes the system will allow
for a single effective uid.
This limit only applies to processes with an effective uid of nonzero
at the time of a fork request.
Processes that have already been started are not affected if the limit
is changed.
.It Li KERN_MAXVNODES
The maximum number of vnodes available on the system.
.It Li KERN_NGROUPS
The maximum number of supplemental groups.
.It Li KERN_NISDOMAINNAME
The name of the current YP/NIS domain.
.It Li KERN_OSRELDATE
The kernel release version in the format
.Ar M Ns Ar mm Ns Ar R Ns Ar xx ,
where
.Ar M
is the major version,
.Ar mm
is the two digit minor version,
.Ar R
is 0 if release branch, otherwise 1,
and
.Ar xx
is updated when the available APIs change.
.Pp
The userland release version is available from
.In osreldate.h ;
parse this file if you need to get the release version of
the currently installed userland.
.It Li KERN_OSRELEASE
The system release string.
.It Li KERN_OSREV
The system revision string.
.It Li KERN_OSTYPE
The system type string.
.It Li KERN_POSIX1
The version of
.St -p1003.1
with which the system
attempts to comply.
.It Li KERN_PROC
Return the entire process table, or a subset of it.
An array of pairs of
.Va struct proc
followed by corresponding
.Va struct eproc
structures is returned,
whose size depends on the current number of such objects in the system.
The third and fourth level names are as follows:
.Bl -column "Third level nameXXXXXX" "Fourth level is:XXXXXX" -offset indent
.It "Third level name	Fourth level is:"
.It "KERN_PROC_ALL	None"
.It "KERN_PROC_PID	A process ID"
.It "KERN_PROC_PGRP	A process group"
.It "KERN_PROC_TTY	A tty device"
.It "KERN_PROC_UID	A user ID"
.It "KERN_PROC_RUID	A real user ID"
.El
.It Li KERN_PROF
Return profiling information about the kernel.
If the kernel is not compiled for profiling,
attempts to retrieve any of the KERN_PROF values will
fail with
.Er ENOENT .
The third level names for the string and integer profiling information
is detailed below.
The changeable column shows whether a process with appropriate
privilege may change the value.
.Bl -column "GPROFXGMONPARAMXXX" "struct gmonparamXXX" -offset indent
.It Sy "Third level name	Type	Changeable"
.It "GPROF_STATE	integer	yes"
.It "GPROF_COUNT	u_short[\|]	yes"
.It "GPROF_FROMS	u_short[\|]	yes"
.It "GPROF_TOS	struct tostruct	yes"
.It "GPROF_GMONPARAM	struct gmonparam	no"
.El
.Pp
The variables are as follows:
.Bl -tag -width 6n
.It Li GPROF_STATE
Returns GMON_PROF_ON or GMON_PROF_OFF to show that profiling
is running or stopped.
.It Li GPROF_COUNT
Array of statistical program counter counts.
.It Li GPROF_FROMS
Array indexed by program counter of call-from points.
.It Li GPROF_TOS
Array of
.Va struct tostruct
describing destination of calls and their counts.
.It Li GPROF_GMONPARAM
Structure giving the sizes of the above arrays.
.El
.It Li KERN_QUANTUM
The maximum period of time, in microseconds, for which a process is allowed
to run without being preempted if other processes are in the run queue.
.It Li KERN_SAVED_IDS
Returns 1 if saved set-group and saved set-user ID is available.
.It Li KERN_SECURELVL
The system security level.
This level may be raised by processes with appropriate privilege.
It may not be lowered.
.It Li KERN_VERSION
The system version string.
.It Li KERN_VNODE
Return the entire vnode table.
Note, the vnode table is not necessarily a consistent snapshot of
the system.
The returned data consists of an array whose size depends on the
current number of such objects in the system.
Each element of the array contains the kernel address of a vnode
.Va struct vnode *
followed by the vnode itself
.Va struct vnode .
.El
.Ss CTL_MACHDEP
The set of variables defined is architecture dependent.
The following variables are defined for the i386 architecture.
.Bl -column "CONSOLE_DEVICEXXX" "struct bootinfoXXX" -offset indent
.It Sy "Second level name	Type	Changeable"
.It Li "CPU_CONSDEV	dev_t	no"
.It Li "CPU_ADJKERNTZ	int	yes"
.It Li "CPU_DISRTCSET	int	yes"
.It Li "CPU_BOOTINFO	struct bootinfo	no"
.It Li "CPU_WALLCLOCK	int	yes"
.El
.Ss CTL_NET
The string and integer information available for the CTL_NET level
is detailed below.
The changeable column shows whether a process with appropriate
privilege may change the value.
.Bl -column "Second level nameXXXXXX" "routing messagesXXX" -offset indent
.It Sy "Second level name	Type	Changeable"
.It "PF_ROUTE	routing messages	no"
.It "PF_INET	IPv4 values	yes"
.It "PF_INET6	IPv6 values	yes"
.El
.Pp
.Bl -tag -width 6n
.It Li PF_ROUTE
Return the entire routing table or a subset of it.
The data is returned as a sequence of routing messages (see
.Xr route 4
for the header file, format and meaning).
The length of each message is contained in the message header.
.Pp
The third level name is a protocol number, which is currently always 0.
The fourth level name is an address family, which may be set to 0 to
select all address families.
The fifth and sixth level names are as follows:
.Bl -column "Fifth level nameXXXXXX" "Sixth level is:XXX" -offset indent
.It Sy "Fifth level name	Sixth level is:"
.It "NET_RT_FLAGS	rtflags"
.It "NET_RT_DUMP	None"
.It "NET_RT_IFLIST	0 or if_index"
.It "NET_RT_IFMALIST	0 or if_index"
.El
.Pp
The
.Dv NET_RT_IFMALIST
name returns information about multicast group memberships on all interfaces
if 0 is specified, or for the interface specified by
.Va if_index .
.It Li PF_INET
Get or set various global information about the IPv4
(Internet Protocol version 4).
The third level name is the protocol.
The fourth level name is the variable name.
The currently defined protocols and names are:
.Bl -column ProtocolXX VariableXX TypeXX ChangeableXX
.It Sy "Protocol	Variable	Type	Changeable"
.It "icmp	bmcastecho	integer	yes"
.It "icmp	maskrepl	integer	yes"
.It "ip	forwarding	integer	yes"
.It "ip	redirect	integer	yes"
.It "ip	ttl	integer	yes"
.It "udp	checksum	integer	yes"
.El
.Pp
The variables are as follows:
.Bl -tag -width 6n
.It Li icmp.bmcastecho
Returns 1 if an ICMP echo request to a broadcast or multicast address is
to be answered.
.It Li icmp.maskrepl
Returns 1 if ICMP network mask requests are to be answered.
.It Li ip.forwarding
Returns 1 when IP forwarding is enabled for the host,
meaning that the host is acting as a router.
.It Li ip.redirect
Returns 1 when ICMP redirects may be sent by the host.
This option is ignored unless the host is routing IP packets,
and should normally be enabled on all systems.
.It Li ip.ttl
The maximum time-to-live (hop count) value for an IP packet sourced by
the system.
This value applies to normal transport protocols, not to ICMP.
.It Li udp.checksum
Returns 1 when UDP checksums are being computed and checked.
Disabling UDP checksums is strongly discouraged.
.Pp
For variables net.inet.*.ipsec, please refer to
.Xr ipsec 4 .
.El
.It Li PF_INET6
Get or set various global information about the IPv6
(Internet Protocol version 6).
The third level name is the protocol.
The fourth level name is the variable name.
.Pp
For variables net.inet6.* please refer to
.Xr inet6 4 .
For variables net.inet6.*.ipsec6, please refer to
.Xr ipsec 4 .
.El
.Ss CTL_USER
The string and integer information available for the CTL_USER level
is detailed below.
The changeable column shows whether a process with appropriate
privilege may change the value.
.Bl -column "USER_COLL_WEIGHTS_MAXXXX" "integerXXX" -offset indent
.It Sy "Second level name	Type	Changeable"
.It "USER_BC_BASE_MAX	integer	no"
.It "USER_BC_DIM_MAX	integer	no"
.It "USER_BC_SCALE_MAX	integer	no"
.It "USER_BC_STRING_MAX	integer	no"
.It "USER_COLL_WEIGHTS_MAX	integer	no"

X11/ChangeLog
2006-09-08  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_44

2006-09-06  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.h,
	xc/lib/GL/apple/dri_dispatch.defs: removed glPointParameteri[v]
	aliases

	* xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/applewmstr.h,
	xc/programs/Xserver/hw/apple/applewm.h,
	xc/programs/Xserver/hw/apple/applewm.c,
	xc/programs/Xserver/hw/apple/applewm-impl.h,
	xc/programs/Xserver/hw/apple/applewm-impl.c,
	xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: use shared
	version of AppleWM headers

	* make-host-def, build.sh, Makefile: removed Xplugin/quartz-wm
	references; attempt to support x86_64

2006-07-07  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_43

	* xc/programs/Xserver/hw/apple/x-hash.c: fix compiler warning

	* xc/programs/xterm/main.c: fix xterm build failure (4607761)

2006-02-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/GL/apple/indirect.c
	(glAquaInitVisualConfigs): add stereo visuals (4377346)

2006-01-16  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_42

	* xc/programs/Xserver/hw/apple/bundle-main.c: added /sbin and
	/usr/sbin to the default PATH value (4408133)

	* xc/programs/Xserver/hw/apple/quartz.c (do_exec): use same
	default PATH and MANPATH values as bundle-main.c (4408133)

2005-11-29  John Harper  <jsh@apple.com>

	Fix xscreensaver-demo crash and make its cross-process GL
	surfaces work correctly (4356702):

	* xc/programs/Xserver/hw/apple/dri-surface.c: don't compile the
	code to draw on surfaces (that was always disables but was
	causing a crash - I've fixed the crash, but the code is never
	used so best to just remove it from the binary)

2005-10-25  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_41

	* xc/lib/xtrans/Xtranssock.c: fix Leopard build failure
	(4222517)

2005-10-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix default
	MANPATH (4312329)

2005-08-29  John Harper  <jsh@apple.com>

	* xc/extras/freetype2/src/base/ftmac.c: merged endian changes
	from 3477131

2005-06-22  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: merged
	fix for CAN-2005-0605 from xfree86 cvs (4031381)

2005-05-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c
	(install_ipaddr_source): changed the way we listen for network
	/ hostname changes, hopefully to be more reliable

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	change fix for bring-all-to-front not to require private
	headers - hardcode enum value we need :-(  (4075093)

2005-05-23  John Harper  <jsh@apple.com>

	* Makefile, make-host-def, merge-libs.sh,
	xc/config/cf/Imake.cf, xc/config/cf/darwin.cf,
	xc/config/cf/darwinLib.tmpl, xc/include/Xmd.h,
	xc/lib/Xt/TMaction.c, xc/lib/Xt/Xtos.h,
	xc/lib/freetype2/Imakefile: checked in changes to
	support building fat libraries when ppc64 is in RC_ARCHS
	(3721701)

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	bumped short version number to 1.2 so we don't fail to verify

	* xc/programs/Xserver/hw/apple/X11Application.m: handle scroll
	wheel events with delta < 1 correctly (4109092)

2005-05-20  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless.h,
	xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m: recognize the
	bfCPSOrderAllWindowsForward bit in activation events and
	respond to it appropriately (4075093)

2005-03-09  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_40

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-03  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_39

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_38

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_37

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-24  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36_1

	* Makefile: strip-libraries target doesn't strip libXplugin

2004-11-23  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)


autofs/driver/sysctl.c
/*
 * Copyright (c) 2003, 2004 Apple Computer, Inc. All rights reserved.
 *
 * Created by Patrick Dirks on Mon Mar 17 2003.
 *
 * $Id: sysctl.c,v 1.2 2004/05/27 08:07:47 pwd Exp $
 */

#include <sys/types.h>
#include <sys/param.h>
#include <sys/errno.h>
#include <sys/mount.h>
#include <sys/sysctl.h>
#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>

#include "sysctl.h"

int
vfsevent_init(void)
{
	int error, kq;
	struct kevent ki;
	struct timespec to;

	bzero(&to, sizeof(to));
	kq = kqueue();
	if (kq == -1)
		err(1, "kqueue");
	EV_SET(&ki, 0, EVFILT_FS, EV_ADD, 0, 0, 0);
again:
	error = kevent(kq, &ki, 1, NULL, 0, &to);
	if (error == -1) {
		if (errno == EINTR)
			goto again;
		err(1, "kevent(EVFILT_FS)");
	}
	return (kq);
}

u_int
vfsevent_wait(int kq, int timeout)
{
	int error;
	struct timespec to, *tp;
	struct kevent ko;

	if (timeout) {
		tp = &to;
		to.tv_sec = timeout;
		to.tv_nsec = 0;
		//dbg("waiting (up to %d seconds) for fs event...\n", timeout);
	} else {
		tp = NULL;
		//dbg("waiting (forever) for fs event...\n");
	}
again:
	error = kevent(kq, NULL, 0, &ko, 1, tp);
	if (error == -1) {
		if (errno == EINTR)
			goto again;
		err(1, "kevent(EVFILT_FS)");
	}
	return (error == 0 ? 0 : ko.fflags);
}

int
sysctl_fsid(op, fsid, oldp, oldlenp, newp, newlen)
	int op;
	fsid_t *fsid;
	void *oldp;
	size_t *oldlenp;
	void *newp;
	size_t newlen;
{
	int ctlname[CTL_MAXNAME+2];
	size_t ctllen;
	const char *sysstr = "vfs.generic.ctlbyfsid";
	struct vfsidctl vc;

	ctllen = CTL_MAXNAME+2;
	if (sysctlnametomib(sysstr, ctlname, &ctllen) == -1) {
		warn("sysctlnametomib(%s)", sysstr);
		return -1;
	};
	ctlname[ctllen] = op;

	bzero(&vc, sizeof(vc));
	vc.vc_vers = VFS_CTL_VERS1;
	vc.vc_fsid = *fsid;
	vc.vc_ptr = newp;
	vc.vc_len = newlen;
	return (sysctl(ctlname, ctllen + 1, oldp, oldlenp, &vc, sizeof(vc)));
}

int
vfslist_get(fsid_t **fsidp, int *count)
{
	int x;
	size_t olen;
	fsid_t *fsidlst;
	const char *sysstr = "vfs.generic.vfsidlist";

rescan:
	x = sysctlbyname(sysstr, NULL, &olen, NULL, 0);
	if (x == -1) return -1;

	fsidlst = malloc(olen);
	if (fsidlst == NULL) return -1;

	x = sysctlbyname(sysstr, fsidlst, &olen, NULL, 0);
	if (x == -1) {
		if (errno == ENOMEM) {
			free(fsidlst);
			goto rescan;
		}
		return -1;
	}
	*fsidp = fsidlst;
	*count = (int)(olen / sizeof(fsid_t));
	return 0;
}

void
vfslist_free(fsid_t *fsid)
{
	free(fsid);
}

int
sysctl_statfs(fsid_t *fsid, struct statfs *sfs, int flag)
{
	size_t size;

	size = sizeof(*sfs);
	return (sysctl_fsid(VFS_CTL_STATFS, fsid, sfs, &size, &flag,
		sizeof(flag)));
}

int
sysctl_queryfs(fsid_t *fsid, struct vfsquery *vq)
{
	size_t size;

	size = sizeof(*vq);
	return (sysctl_fsid(VFS_CTL_QUERY, fsid, vq, &size, NULL, 0));
}

int
sysctl_unmount(fsid_t *fsid, int flag)
{

	return (sysctl_fsid(VFS_CTL_UMOUNT, fsid, NULL, 0, &flag,
		sizeof(flag)));
}

int
sysctl_setfstimeout(fsid_t *fsid, int timeo)
{

	return (sysctl_fsid(VFS_CTL_TIMEO, fsid, NULL, 0,
		    &timeo, sizeof(timeo)));
}

int
sysctl_getfstimeout(fsid_t *fsid, int *timeop)
{
	size_t size = sizeof(*timeop);

	return (sysctl_fsid(VFS_CTL_TIMEO, fsid, timeop, &size,
		    NULL, 0));
}
